runtime/ts/storage/firebase-storage.ts

```diff
@@ -397,67 +473,93 @@ class FirebaseVariable extends FirebaseStorageProvider {
 
   async get() {
     await this.initialized;
-    if (this.referenceMode) {
+    if (this.referenceMode && this.value) {
       const referredType = this.type;
-      if (this.backingStore == null) {
-        const backingStore = await this.storageEngine.share(
-          referredType.toString(),
-          referredType.collectionOf(),
-          this.value.storageKey);
-        this.backingStore = backingStore as FirebaseCollection;
-      }
+      await this.ensureBackingStore();
       return await this.backingStore.get(this.value.id);  
     }
     return this.value;
   }
 
   async set(value, originatorId=null, barrier=null) {
-    let referredType;  
-    // the await required for fetching baseStorage can cause initialization/localModified
-    // flag reordering if done inline below. So we resolve backingStore if necessary
-    // first, before looking at anything else. 
-    if (this.referenceMode && this.backingStore == null) {
-      referredType = this.type;    
-      this.backingStore = await this.storageEngine.baseStorageFor(referredType, this.storageEngine.baseStorageKey(referredType, this.storageKey));
-    }
-
+    assert(value !== undefined);
     if (this.version == null) {
       assert(!this.localModified);
       // If the first modification happens before init, this becomes
       // init. We pick the initial version which will be updated by the
       // transaction in _persistChanges.
       this.version = 0;
       this.resolveInitialized();
-    } else {
+    } else if (!this.referenceMode) {
+      // If in reference mode, we can't actually determine if this value is identical to the previous
+      // one.
       if (JSON.stringify(this.value) === JSON.stringify(value)) {
          return;
       }
-      this.version++;
     }
-    if (this.referenceMode) {
-      await this.backingStore.store(value, [this.storageKey]);
-      value = {id: value.id, storageKey: this.backingStore.storageKey};
+    this.version++;
+    const version = this.version;
+    let storageKey;
+    if (this.referenceMode && value) {
+      storageKey = this.storageEngine.baseStorageKey(this.type, this.storageKey);
+      this.value = {id: value.id, storageKey};
+      this.pendingWrites.push({value, storageKey});
+    } else {
+      this.value = value;
     }
-
     this.localModified = true;
-    this.value = value;
-    this._fire('change', {data: this.value, version: this.version, originatorId, barrier});
-    await this._persistChanges();
+
+    // the await required for fetching baseStorage can cause initialization/localModified
+    // flag reordering if done before persisting a change.
```

Comment:
..., so we need to fire the change event after persistChanges has finished?

Reply:
oops, out-of-date comment.